src/core/transformScriptSetup.ts

```diff
@@ -1,228 +1,231 @@
-    hasBody = true
-    const directivesObject = t.objectExpression(
-      directiveDeclaration.map(({ directive, identifier }) =>
-        t.objectProperty(
-          t.identifier(directive),
-          t.identifier(identifier),
-          false,
-          false,
-        ),
-      ),
-    )
-
-    ast.body.push(
-      t.expressionStatement(
-        t.assignmentExpression(
-          '=',
-          t.memberExpression(__sfc, t.identifier('directives')),
-          t.callExpression(
-            t.memberExpression(t.identifier('Object'), t.identifier('assign')),
-            [
-              directivesObject,
-              t.memberExpression(__sfc, t.identifier('directives')),
-            ],
-          ),
-        ),
-      ) as any,
-    )
-  }
-
-  if (!hasBody && !options?.astTransforms) {
-    return {
-      ast: null,
-      code: '',
-    }
-  }
-
-  // re-export
-  // `export default __sfc_main`
-  ast.body.push(t.exportDefaultDeclaration(__sfc) as any)
-
-  ast = options?.astTransforms?.post?.(ast, sfc) || ast
-
-  return {
-    ast,
-    code: generate(ast).code,
-  }
-}
+import { capitalize } from '@vue/shared'
+import type { Node, ObjectExpression, Statement } from '@babel/types'
+import { notNullish, partition, uniq } from '@antfu/utils'
+import { parserOptions } from '@vue/compiler-dom'
+import type { ParsedSFC, ScriptSetupTransformOptions } from '../types'
+import { applyMacros } from './macros'
+import { getIdentifierDeclarations } from './identifiers'
+import { generate, t } from './babel'
+import { pascalize } from './utils'
+
+function isAsyncImport(node: Statement) {
+  if (t.isVariableDeclaration(node)) {
+    const declaration = node.declarations[0]
+
+    return (
+      declaration !== undefined
+      && t.isCallExpression(declaration.init)
+      && t.isIdentifier(declaration.init.callee)
+      && declaration.init.callee.name === 'defineAsyncComponent'
+    )
+  }
+
+  return false
+}
+
+export function transformScriptSetup(
+  sfc: ParsedSFC,
+  options?: ScriptSetupTransformOptions,
+) {
+  const { scriptSetup, script, template } = sfc
+
+  const { nodes: body, props, expose } = applyMacros(scriptSetup.ast.body)
+
+  const [hoisted, setupBody] = partition(
+    body,
+    n =>
+      isAsyncImport(n)
+      || t.isImportDeclaration(n)
+      || t.isExportNamedDeclaration(n)
+      || n.type.startsWith('TS'),
+  )
+
+  // get all identifiers in `<script setup>` and `<script>`
+  const declarationArray = uniq([
+    ...getIdentifierDeclarations(hoisted),
+    ...getIdentifierDeclarations(setupBody),
+    ...getIdentifierDeclarations(script.ast.body),
+  ]).filter(notNullish)
+
+  // filter out identifiers that are used in `<template>`
+  const returns: ObjectExpression['properties'] = declarationArray
+    .filter(i => template.identifiers.has(i))
+    .map((i) => {
+      const id = t.identifier(i)
+      return t.objectProperty(id, id, false, true)
+    })
+
+  const components = Array.from(template.tags)

```

Comment:
What would here happen if `template.tags` is `['hello-world','HelloWorld']`

Reply:
> What would here happen if `template.tags` is `['hello-world','HelloWorld']`
I don't filter these non-native tags,The result should be the same as before

```js
//template.tags is ['hello-world','HelloWorld']
 const nonNativeTags = new Set(
    Array.from(template.tags)
      .filter(tag => !parserOptions.isNativeTag!(tag))
      .map(pascalize),
  )
  
console.log(nonNativeTags)//['hello-world','HelloWorld']

 const components = Array.from(nonNativeTags)
    .map(
      component =>
        declarationArray.find(declare => declare === component)
        ?? declarationArray.find(declare => pascalize(declare) === component),
    )
    .filter(notNullish)
```